lib/Internal/DisposableStream.php

```diff
@@ -0,0 +1,43 @@
+<?php
+
+namespace Amp\Internal;
+
+use Amp\Promise;
+use Amp\Stream;
+
+/**
+ * Wraps a Stream instance that has public methods to yield, complete, and fail into an object that only allows
+ * access to the public API methods and sets $disposed to true when the object is destroyed.
+ *
+ * @internal
+ *
+ * @template-covariant TValue
+ * @template-implements Stream<TValue>
+ */
+class DisposableStream implements Stream
```

Comment:
I'd expect a `dispose()` method on a `DisposableStream`

Reply:
Maybe call it `DestructibleStream` then?

Reply:
No, I'd offer a `dispose()` method, it can be useful if you have a circular deference and don't want to call `gc_collect_cycles()`.

Reply:
Added.